File: bfs_graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution {
public:
	vector<int> bfsOfGraph(int V, vector<int> adj[]){
	    // Code here
	    queue<int> q;
	    vector<int> visited(V);
	    vector<int> res;
	    q.push(0);
	    while(!q.empty()) {
	        int c = q.front();
	        q.pop();
	        res.push_back(c);
	        visited[c] = 1;
	        for(int i = 0; i < adj[c].size(); ++i) {
	            if (!visited[adj[c][i]]) {
	                q.push(adj[c][i]);
	                visited[adj[c][i]] = 1;
	            }
	        }
	    }
	    return res;
	}
};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int V, E;
        cin >> V >>

            E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
            // 		adj[v].push_back(u);
        }
        // string s1;
        // cin>>s1;
        Solution obj;
        vector<int> ans = obj.bfsOfGraph(V, adj);
        for (int i = 0; i < ans.size(); i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: graph/bfs_graph.py
# User function Template for python3

from typing import List
from queue import Queue


class Solution:
    # Function to return Breadth First Traversal of given graph.
    def bfsOfGraph(self, V: int, adj: List[List[int]]) -> List[int]:
        # code here
        q = Queue()
        q.put(0)
        marked = [0] * V
        marked[0] = 1
        ans = []
        while not q.empty():
            v = q.get()
            ans.append(v)
            for w in adj[v]:
                if marked[w] == 0:
                    marked[w] = 1
                    q.put(w)

        return ans


# {
# Driver Code Starts


if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        V, E = map(int, input().split())
        adj = [[] for i in range(V)]
        for _ in range(E):
            u, v = map(int, input().split())
            adj[u].append(v)
        ob = Solution()
        ans = ob.bfsOfGraph(V, adj)
        for i in range(len(ans)):
            print(ans[i], end=" ")
        print()


# } Driver Code Ends
